### ISSUE TITLE
Use waitRules to wait for custom resource

### ISSUE THREAD

**Issue by johny-mnemonic** [k14s/kapp#192]
I have tried to find more examples of using waitRules, but even mighty Google is silent, so I would like to ask for help here.
I am delivering CRD for Kubernetes Operator (more specifically MinIO Operaror) by Kapp, and would like to convince Kapp to wait for the object created by the Operator.

I am adding this Kapp config to the list of CRDs, but it doesn't seems to work:
```
---
apiVersion: kapp.k14s.io/v1alpha1
kind: Config
metadata:
  name: kapp-config

minimumRequiredVersion: 0.29.0

waitRules:
- supportsObservedGeneration: true
  conditionMatchers:
  - type: Ready
    status: "False"
    failure: true
  - type: Ready
    status: "True"
    success: true
  resourceMatchers:
  - kindNamespaceNameMatcher: {kind: Pod, namespace: minio, name: minio-zone-0-0}
```

Any hint what I might be doing wrong would help.
I hope that this is not some antipattern. With recent sprawl of operators, I guess it will be pretty common to still deliver the app by Kapp, but with the operator, it means delivering only it's definition/config for the operator, not the actual app itself.

Edit: It would be even better to wait for the minio service same way as Kapp normally do (i.e. for /v1/Service kind wait for spec.clusterIP). Which means I don't even need to define custom wait rules, just a custom resource to wait for.

**Comment by joaopapereira** [comment]
Hello @johny-mnemonic 
the `waitRules` configuration rules concept was created to enable `kapp` to wait for a particular resource that it is creating to be in a particular state. In your case looks like that you are creating a Custom Resource that will be consumed by an Operator and that Operator would create other resources in the cluster.
In this pattern, I am not sure if `kapp` should be aware of these other Resources that `kapp` will not manage directly. 

But in your case, it feels like you should have an option to wait on the Custom Resource that you deploy with `kapp`. If the Operator would update the status of the Custom Resource you are deploying with `kapp` you could wait on that status. Do you think this would be possible? Do you have an example of the Custom Resource that you are using?

**Comment by johny-mnemonic** [comment]
Thanks for reply @joaopapereira
That's correct, I am deploying custom resource that is consumed by an Operator, but unfortunately that resource doesn't have status, so there is nothing Kapp could wait for.
Sure, Kapp shouldn't be aware of resources created by something else (like an Operator) by default, but I hoped there is a way to convince Kapp to wait for custom defined resource in case you need to.

Your suggestion, that the custom resource should have status updated by the Operator is probably the best solution of my issue, but I am not a developer of this Operator. Will try to ask the developers whether they like the idea and maybe they will implement it. 
I am deploying MinIO [tenant definition](https://github.com/minio/operator/blob/master/examples/tenant.yaml#L31) by Kapp.

**Comment by joaopapereira** [comment]
I did dig a little bit around the operator code and looks like they provide some sort of status but it is not part of the conditions. Which is unfortunate. Nevertheless, this should be a pretty easy change on the operator and since the majority of the CRDs around the k8s ecosystem use conditions to communicate the status, it would be a nice addition to their operator.

**Comment by StevenLocke** [comment]
With some looking around, we've added this issue to the "Cluster Fulfilled Resources" epic. The idea there is to give kapp some notion of resources that it won't directly create, but will be accessible. When kapp is able to "know" about those resources, it should also know how to wait for them, which should resolve this request.

This also feels similar to #11, which is still being discussed.

**Comment by cari-lynn** [comment]
@johny-mnemonic Is this still an issue for you? This issue is very similar to the issue @StevenLocke linked to above #11, there is some discussion happening in that issue around ways to solve this. Does that other issue capture your problem? If so I would like to mark this as a duplicate and continue discussion in the other issue.

**Comment by johny-mnemonic** [comment]
@cari-lynn yes #11 looks very similar, but I am not sure what the dollar sign in "$.status" stands for.
In case it stands for the custom resource delivered by Kapp and is limited to it, then my request is maybe a bit more generic.
I wanted to be able to wait for any resource in the cluster.

**Comment by praveenrewar** [comment]
We have added support of [kapp.k14s.io/exists](https://carvel.dev/kapp/docs/latest/apply/#kappk14sioexists-available-in-v0430) annotation (available in v0.43.0) which can be used to wait for resources that are not owned by kapp.
